pkg/controllers/disruption/queue.go

```diff
@@ -282,44 +216,135 @@ func (q *Queue) waitOrTerminate(ctx context.Context, cmd *Command) error {
+			errs[i] = client.IgnoreNotFound(err)
+			return
+		}
 	})
+	var markedCandidates []*Candidate
+	for i := range errs {
+		if errs[i] != nil {
+			continue
+		}
+		markedCandidates = append(markedCandidates, cmd.Candidates[i])
+	}
+	return markedCandidates, multierr.Combine(errs...)
+}
+
+// createReplacementNodeClaims creates replacement NodeClaims
+func (q *Queue) createReplacementNodeClaims(ctx context.Context, cmd *Command) error {
+	nodeClaimNames, err := q.provisioner.CreateNodeClaims(ctx, lo.Map(cmd.Replacements, func(r *Replacement, _ int) *pscheduling.NodeClaim { return r.NodeClaim }), provisioning.WithReason(strings.ToLower(string(cmd.Reason()))))
+	if err != nil {
+		return err
+	}
+	if len(nodeClaimNames) != len(cmd.Replacements) {
+		// shouldn't ever occur since a partially failed CreateNodeClaims should return an error
+		return serrors.Wrap(fmt.Errorf("expected replacement count did not equal actual replacement count"), "expected-count", len(cmd.Replacements), "actual-count", len(nodeClaimNames))
+	}
+	for i, name := range nodeClaimNames {
+		cmd.Replacements[i].Name = name
+	}
+	return nil
+}
+
+// StartCommand will do the following:
+// 1. Taint candidate nodes
+// 2. Spin up replacement nodes
+// 3. Add Command to the queue to wait to delete the candidates.
+func (q *Queue) StartCommand(ctx context.Context, cmd *Command) error {
 	// First check if we can add the command.
-	if q.HasAny(providerIDs...) {
+	if q.HasAny(lo.Map(cmd.Candidates, func(c *Candidate, _ int) string { return c.ProviderID() })...) {
 		return fmt.Errorf("candidate is being disrupted")
 	}
 
-	cmd.timeAdded = q.clock.Now()
+	log.FromContext(ctx).WithValues(append([]any{"command-id", cmd.ID, "reason", strings.ToLower(string(cmd.Reason()))}, cmd.LogValues()...)...).Info("disrupting node(s)")
+
+	// Cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	markedCandidates, markDisruptedErr := q.markDisrupted(ctx, cmd)
+	// If we get a failure marking some nodes as disrupted, if we are launching replacements, we shouldn't continue
+	// with disrupting the candidates. If it's just a delete operation, we can proceed
+	if markDisruptedErr != nil && len(cmd.Replacements) > 0 {
+		return serrors.Wrap(fmt.Errorf("marking disrupted, %w", markDisruptedErr), "command-id", cmd.ID)
+	}
+
+	// Update the command to only consider the successfully MarkDisrupted candidates
+	cmd.Candidates = markedCandidates
+
+	if err := q.createReplacementNodeClaims(ctx, cmd); err != nil {
+		// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
+		// we don't want to disrupt workloads with no way to provision new nodes for them.
+		return serrors.Wrap(fmt.Errorf("launching replacement nodeclaim, %w", err), "command-id", cmd.ID)
```

Comment:
When this fails, the replacement nodeclaims that were created will be garbage collected because there is no node for the nodeclaim?

Reply:
Not sure that I follow -- if a portion of the replacements fails to come up, then there will be a few more NodeClaims present than there should be -- this is part of the failure mode that we expect to clean-up through emptiness. There really isn't a better way to handle this than that since this is a declarative system and any imperative action that we would attempt to take could fail at any poitn